lib/cloud/gcp/vm.go

```diff
@@ -0,0 +1,545 @@
+}
+
+// CheckAndSetsDefaults checks and sets defaults for InstancesClientConfig.
+func (c *InstancesClientConfig) CheckAndSetDefaults(ctx context.Context) (err error) {
+	if c.InstanceClient == nil {
+		if c.InstanceClient, err = compute.NewInstancesRESTClient(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// Instance represents a GCP VM.
+type Instance struct {
+	// Name is the instance's name.
+	Name string
+	// Zone is the instance's zone.
+	Zone string
+	// ProjectID is the ID of the project the VM is in.
+	ProjectID string
+	// ServiceAccount is the email address of the VM's service account, if any.
+	ServiceAccount string
+	// Labels is the instance's labels.
+	Labels map[string]string
+
+	ipAddress string
+	hostKeys  []ssh.PublicKey
+	metadata  *computepb.Metadata
+}
+
+// InstanceRequest formats an instance request based on an instance.
+func (i *Instance) InstanceRequest() InstanceRequest {
+	return InstanceRequest{
+		ProjectID: i.ProjectID,
+		Zone:      i.Zone,
+		Name:      i.Name,
+	}
+}
+
+// NewInstanccesClient creates a new InstancesClient.
+func NewInstancesClient(ctx context.Context) (InstancesClient, error) {
+	var cfg InstancesClientConfig
+	client, err := NewInstancesClientWithConfig(ctx, cfg)
+	return client, trace.Wrap(err)
+}
+
+// NewInstanccesClientWithConfig creates a new InstancesClient with custom
+// config.
+func NewInstancesClientWithConfig(ctx context.Context, cfg InstancesClientConfig) (InstancesClient, error) {
+	if err := cfg.CheckAndSetDefaults(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &instancesClient{InstancesClientConfig: cfg}, nil
+}
+
+// instancesClient implements the InstancesClient interface by wrapping
+// compute.InstancesClient.
+type instancesClient struct {
+	InstancesClientConfig
+}
+
+func isExternalNAT(s string) bool {
+	return slices.Contains([]string{
+		"external-nat",
+		"External NAT",
+	}, s)
+}
+
+func toInstance(origInstance *computepb.Instance) *Instance {
+	zoneParts := strings.Split(origInstance.GetZone(), "/")
+	zone := zoneParts[len(zoneParts)-1]
+	var ipAddress string
+outer:
+	for _, netInterface := range origInstance.GetNetworkInterfaces() {
+		for _, accessConfig := range netInterface.GetAccessConfigs() {
+			if isExternalNAT(accessConfig.GetName()) {
+				ipAddress = accessConfig.GetNatIP()
+				break outer
+			}
+		}
+	}
+	inst := &Instance{
+		Name:      origInstance.GetName(),
+		Zone:      zone,
+		Labels:    origInstance.GetLabels(),
+		ipAddress: ipAddress,
+		metadata:  origInstance.GetMetadata(),
+	}
+	// GCP VMs can have at most one service account.
+	if len(origInstance.ServiceAccounts) > 0 {
+		inst.ServiceAccount = origInstance.ServiceAccounts[0].GetEmail()
+	}
+	return inst
+}
+
+// ListInstances lists the GCP VMs that belong to the given project and
+// zone.
+// zone supports wildcard "*".
+func (clt *instancesClient) ListInstances(ctx context.Context, projectID, zone string) ([]*Instance, error) {
```

Comment:
TBH I think that instead of hardcoding pagination into this method and always trying to fetch the entire list of VMs, it should accept the nextPageToken returned by GCP API to allow the caller to control pagination.

I'm just thinking of a scenario where, say, there are thousands of instances and we might want to throttle the API calls, with this API interface it's not possible for the caller to control.

You can still have a separate ListInstances function that would accept this client and paginate over everything as a convenient helper method, but I think that in the client API we should let the caller control the pagination.

Reply:
I've added `StreamInstances` using `stream.Stream` so the caller doesn't have to worry about the next page token. Let me know if that works for you or if you want to paginate differently.

Reply:
the stream.Steam is used only internally and the api call still doesn't use pageToken so still the 
```go
		it := clt.InstanceClient.AggregatedList(ctx, &computepb.AggregatedListInstancesRequest{
			Project: projectID,
		})
```

will allow to iterate only on first 500 records.

Reply:
I'm reading through the [iterator docs](https://pkg.go.dev/google.golang.org/api/iterator) and I don't think this is actually the case. The docs of [Pager.NextPage](https://pkg.go.dev/google.golang.org/api/iterator#Pager.NextPage), which uses the underlying iterator's page token, say you only need the next page token if you need to resume iteration later. The iterator should be taking care of the page tokens under the hood.